projectile.el

```diff
@@ -2763,35 +2776,44 @@ test/impl/other files as below:
     a plist containing :test, :impl or :other as key and the relative path/paths or
     predicate as value.  PREDICATE accepts a relative path as the input."
-  (if-let ((existing-project-plist
-            (cl-find-if
-             (lambda (p) (eq project-type (car p))) projectile-project-types))
+    (let* ((existing-project-plist
+            (or (cl-find-if
+                 (lambda (p) (eq project-type (car p))) projectile-project-types)
+                (error "No existing project found for: %s" project-type)))
            (new-plist
-            (projectile--build-project-plist
-             marker-files
-             :project-file project-file
-             :compilation-dir compilation-dir
-             :configure configure
-             :compile compile
-             :install install
-             :package package
-             :test test
-             :run run
-             :test-suffix test-suffix
-             :test-prefix test-prefix
-             :src-dir src-dir
-             :test-dir test-dir
-             :related-files-fn related-files-fn))
+            (append
+             (when marker-files-specified `(marker-files ,marker-files))
+             (when project-file-specified `(project-file ,project-file))
+             (when compilation-dir-specified `(compilation-dir ,compilation-dir))
+             (when configure-specified `(configure-command ,configure))
+             (when compile-specified `(compile-command ,compile))
+             (when test-specified `(test-command ,test))
+             (when install-specified `(install-command ,install))
+             (when package-specified `(package-command ,package))
+             (when run-specified `(run-command ,run))
+             (when test-suffix-specified `(test-suffix ,test-suffix))
+             (when test-prefix-specified `(test-prefix ,test-prefix))
+             (when src-dir-specified `(src-dir ,src-dir))
+             (when test-dir-specified `(test-dir ,test-dir))
+             (when related-files-fn-specified
+               `(related-files-fn ,related-files-fn))))
            (merged-plist
             (projectile--combine-plists
              (cdr existing-project-plist) new-plist))
            (project-type-elt (cons project-type merged-plist)))
-      (setq projectile-project-types
-            (mapcar (lambda (p) (if (eq project-type (car p))
-                                    project-type-elt
-                                  p))
-                    projectile-project-types))
-    (error "No existing project found for: %s" project-type)))
+      (cl-flet* ((project-filter (p) (eq project-type (car p)))
+                 (project-map (p) (if (project-filter p) project-type-elt p)))
+        (setq projectile-project-types
+              (if precedence
+                  (let ((filtered-types
+                       (cl-remove-if #'project-filter projectile-project-types)))
+                    (setq projectile-project-type-cache (make-hash-table))
```

Comment:
Invalidate the project type cache here, I think this is what we would want since otherwise users may have to remove projects from `projectile-project-type-cache` manually.